File: 2pc-vote/participant.py
import os
import grpc
from concurrent import futures
import time
from proto import transaction_pb2, transaction_pb2_grpc

class ParticipantServicer(transaction_pb2_grpc.ParticipantServicer):
    def __init__(self):
        self.transactions = {}

    def Prepare(self, request, context):
        print(f"[Participant] PREPARE received for tx {request.transaction_id}")
        self.transactions[request.transaction_id] = "PREPARED"
        return transaction_pb2.Response(transaction_id=request.transaction_id, status="READY")

    def Commit(self, request, context):
        print(f"[Participant] COMMIT received for tx {request.transaction_id}")
        self.transactions[request.transaction_id] = "COMMITTED"
        return transaction_pb2.Response(transaction_id=request.transaction_id, status="COMMITTED")

    def Abort(self, request, context):
        print(f"[Participant] ABORT received for tx {request.transaction_id}")
        self.transactions[request.transaction_id] = "ABORTED"
        return transaction_pb2.Response(transaction_id=request.transaction_id, status="ABORTED")

def join_coordinator(participant_address, retry_interval=5):
    coordinator_address = os.environ.get("COORDINATOR_ADDRESS", "localhost:8080")

    while True:
        print(f"[Participant] Attempting to join coordinator at {coordinator_address} as {participant_address}")
        try:
            channel = grpc.insecure_channel(coordinator_address)
            stub = transaction_pb2_grpc.CoordinatorStub(channel)
            response = stub.Join(transaction_pb2.JoinRequest(address=participant_address))
            print(f"[Participant] Successfully joined coordinator: {response.status}")
            break
        except grpc.RpcError as e:
            print(f"[Participant] Failed to join coordinator: {e}. Retrying in {retry_interval} seconds...")
            time.sleep(retry_interval)

def serve():
    port = os.environ.get("PORT", "8081")
    participant_address = f"{os.environ.get('HOSTNAME', 'participant')}:{port}"

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=5))
    transaction_pb2_grpc.add_ParticipantServicer_to_server(ParticipantServicer(), server)
    server.add_insecure_port(f'[::]:{port}')
    server.start()
    print(f"[Participant] Running on port {port}")

    # Give server a sec to start before joining
    time.sleep(1)
    join_coordinator(participant_address)

    server.wait_for_termination()

if __name__ == '__main__':
    serve()

File: 2pc-vote/javascript/participant.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const sqlite3 = require('sqlite3').verbose();
const fs = require('fs');
const path = require('path');

const PROTO_PATH = path.join(__dirname, 'proto', 'transaction.proto');
const DB_FILE = path.join(__dirname, 'db', 'participant.db');

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    defaults: true,
    enums: String,
    oneofs: true,
});
const proto = grpc.loadPackageDefinition(packageDefinition).proto.transaction;

// Initialize SQLite
const db = new sqlite3.Database(DB_FILE);
db.run(`CREATE TABLE IF NOT EXISTS kv_store (key TEXT PRIMARY KEY, value TEXT)`);

function saveKvPair(key, value) {
    return new Promise((resolve, reject) => {
        db.run(
            `INSERT OR REPLACE INTO kv_store (key, value) VALUES (?, ?)`,
            [key, value],
            function (err) {
                if (err) {
                    console.log(`[Participant] Error saving key-value pair (${key}, ${value}): ${err.message}`);
                    reject(err);
                } else {
                    resolve();
                }
            }
        );
    });
}

const ParticipantService = {
    Prepare: (call, callback) => {
        console.log(`[Participant] PREPARE received for tx ${call.request.transaction_id}`);
        callback(null, {
            transaction_id: call.request.transaction_id,
            status: "READY",
        });
    },

    Commit: async (call, callback) => {
        console.log(`[Participant] COMMIT received for tx ${call.request.transaction_id}`);
        try {
            const data = JSON.parse(call.request.payload);
            const { key, value } = data;
            await saveKvPair(key, value);
            callback(null, {
                transaction_id: call.request.transaction_id,
                status: "COMMITTED",
            });
        } catch (err) {
            console.log(`[Participant] Invalid payload in COMMIT: ${err.message}`);
            callback(null, {
                transaction_id: call.request.transaction_id,
                status: "FAILED",
            });
        }
    },

    Abort: (call, callback) => {
        console.log(`[Participant] ABORT received for tx ${call.request.transaction_id}`);
        callback(null, {
            transaction_id: call.request.transaction_id,
            status: "ABORTED",
        });
    },
};

function joinCoordinator(participantAddress, retryInterval = 5000) {
    const coordinatorAddress = process.env.COORDINATOR_ADDRESS || 'localhost:8080';
    const stub = new proto.Coordinator(coordinatorAddress, grpc.credentials.createInsecure());

    function attemptJoin() {
        console.log(`[Participant] Attempting to join coordinator at ${coordinatorAddress} as ${participantAddress}`);
        stub.Join({ address: participantAddress }, (err, response) => {
            if (err) {
                console.log(`[Participant] Failed to join coordinator: ${err.message}. Retrying in ${retryInterval / 1000}s...`);
                setTimeout(attemptJoin, retryInterval);
            } else {
                console.log(`[Participant] Successfully joined coordinator: ${response.status}`);
            }
        });
    }

    attemptJoin();
}

function main() {
    const port = process.env.PORT || '8081';
    const participantAddress = `${process.env.HOSTNAME || 'localhost'}:${port}`;

    const server = new grpc.Server();
    server.addService(proto.Participant.service, ParticipantService);

    server.bindAsync(
        `0.0.0.0:${port}`,
        grpc.ServerCredentials.createInsecure(),
        (err, port) => {
            if (err) {
                console.error(`[Participant] Server error: ${err.message}`);
                return;
            }
            console.log(`[Participant] Running on port ${port}`);
            setTimeout(() => joinCoordinator(participantAddress), 1000);
        }
    );
}

main();
